core/src/raw/oio/write/chunked_write.rs

```diff
@@ -15,43 +15,87 @@
 // specific language governing permissions and limitations
 // under the License.
 
+use bytes::Buf;
+
 use crate::raw::*;
 use crate::*;
 
 /// ChunkedWriter is used to implement [`oio::Write`] based on chunk:
 /// flush the underlying storage at the `chunk`` size.
 ///
 /// ChunkedWriter makes sure that the size of the data written to the
-/// underlying storage is exactly `chunk` bytes.
+/// underlying storage is
+/// - exactly `chunk` bytes if `exact` is true
+/// - at least `chunk` bytes if `exact` is false
 pub struct ChunkedWriter<W: oio::Write> {
     inner: W,
 
     /// The size for buffer, we will flush the underlying storage at the size of this buffer.
     chunk_size: usize,
+    /// If `exact` is true, the size of the data written to the underlying storage is
+    /// exactly `chunk_size` bytes.
+    exact: bool,
     buffer: oio::QueueBuf,
 }
 
 impl<W: oio::Write> ChunkedWriter<W> {
     /// Create a new exact buf writer.
-    pub fn new(inner: W, chunk_size: usize) -> Self {
+    pub fn new(inner: W, chunk_size: usize, exact: bool) -> Self {
         Self {
             inner,
             chunk_size,
+            exact,
             buffer: oio::QueueBuf::new(),
         }
     }
 }
 
 impl<W: oio::Write> oio::Write for ChunkedWriter<W> {
     async fn write(&mut self, mut bs: Buffer) -> Result<usize> {
-        if self.buffer.len() >= self.chunk_size {
-            let written = self.inner.write(self.buffer.clone().collect()).await?;
+        if self.exact {
+            if self.buffer.len() >= self.chunk_size {
+                let written = self.inner.write(self.buffer.clone().collect()).await?;
+                self.buffer.advance(written);
+            }
+
+            let remaining = self.chunk_size - self.buffer.len();
+            bs.truncate(remaining);
+            let n = bs.len();
+            self.buffer.push(bs);
+            return Ok(n);
+        }
+        // We are in inexact mode.
+
+        if self.buffer.len() + bs.len() < self.chunk_size {
+            // We haven't buffered enough data.
+            let n = bs.len();
+            self.buffer.push(bs);
+            return Ok(n);
+        }
+        // We have enough data to send.
+
+        if self.buffer.is_empty() {
+            // Fast path: Sends the buffer directly if the buffer queue is empty.
+            return self.inner.write(bs).await;
+        }
+
+        // If we always push `bs` to the buffer queue, the buffer queue may grow infinitely if inner
+        // doesn't fully consume the queue. So we clone the buffer queue and send it with `bs` first.
+        let mut buffer = self.buffer.clone();
+        buffer.push(bs.clone());
+        // TODO(yingwen): don't clone buffer.
+        let written = self.inner.write(buffer.clone().collect()).await?;
+        if written < self.buffer.len() {
             self.buffer.advance(written);
+            // We didn't sent `bs`.
+            return Ok(0);
```

Comment:
Write returning `0` often been treated as an error [WriteZero](https://doc.rust-lang.org/std/io/enum.ErrorKind.html#variant.WriteZero). It's better not to do this.

We can push only 1B in the buffer.

This case is extremely unusual, and almost no other service behaves this way. Those rules are just a safety measure. We can make it clear in the comments.

Reply:
What about pushing `written` bytes to the buffer queue? We already sent `written` bytes so the queue should have "written" bytes space left.

Reply:
Makes sense to me.